### DISCUSSION TITLE
Problem compiling

### DISCUSSION THREAD

**Discussion by dukester** [modula3/cm3#1157] [General]
in a <proj-name>/src directory, I have:
MyModule.i3
MyModule.m3
Main.m3
In Main.m3 I have IMPORTS MyModule;
I run cm3 from <proj-name> directory
I'm doing something wrong!

```
--- building in AMD64_LINUX ---
new source -> compiling Main.m3
"../src/Main.m3", line 4: unable to find interface (MyModule)
1 error encountered
m3front failed compiling: ../src/Main.m3
```
```
cat src/m3makefile 
import("libm3") 
implementation("Main") 
program("myModule") 
```

**Comment by olaf-wagner** [comment]
On 26 Feb 2023, at 20:00, Duke Normandin wrote:

> in a <proj-name>/src directory, I have:
> MyModule.i3
> MyModule.m3
> Main.m3
> In Main.m3 I have IMPORTS MyModule;
> I run cm3 from <proj-name> directory
> I'm doing something wrong!

You’re missing the line declaring your module:

module(“MyModule”)

> ```
> --- building in AMD64_LINUX ---
> new source -> compiling Main.m3
> "../src/Main.m3", line 4: unable to find interface (MyModule)
> 1 error encountered
> m3front failed compiling: ../src/Main.m3
> ```
> ```
> cat src/m3makefile
> import("libm3")
> implementation("Main")
> program("myModule")
> ```
>
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/modula3/cm3/discussions/1157
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>




-- 
Olaf Wagner -- elego Software Solutions GmbH -- http://www.elegosoft.com
               Gustav-Meyer-Allee 25 / Gebäude 12, 13355 Berlin, Germany
phone: +49 30 23 45 86 96  mobile: +49 177 2345 869  fax: +49 30 23 45 86 95
Geschäftsführer: Olaf Wagner | Sitz: Berlin
Handelregister: Amtsgericht Charlottenburg HRB 77719 | USt-IdNr: DE163214194

**Reply by dukester** [reply]
Hello Olaf!
module("MyModule") should appear in m3makefile I'm guessing. Is that correct?
Thanks for your input!

**Reply by dukester** [reply]
OK I see! I didn't click on the "..." to see the rest of your reply. Thanks!!

**Comment by dukester** [comment]
Making headway - but not there yet.
```
"../src/Main.m3", line 11: Expression is not a statement
"../src/Main.m3", line 12: parameter not specified (2.3.2):  (r)
```
```
MODULE Main;
(* IMPORT IO;*)
IMPORT MyModule;
VAR firstName: TEXT := "Duke";
BEGIN
  MyModule.MakeThing(firstName, 76);
  MyModule.PrintThing();
END Main.
```
```
INTERFACE MyModule;
CONST MaxEntries = 53;
TYPE
  Thing = RECORD
            name: TEXT;
            size: INTEGER
           END;
PROCEDURE PrintThing(READONLY r: Thing);
PROCEDURE MakeThing(n: TEXT; s: INTEGER): Thing;
END MyModule.
```
Do you need to see the implementation module?

**Comment by mikanystrom** [comment]
change

 MyModule.MakeThing(firstName, 76);
  MyModule.PrintThing();


to



WITH thing = MyModule.MakeThing(firstName, 76) DO
  MyModule.PrintThing(thing)

END


(you need to pass the result of the MakeThing to the PrintThing)


Alternatively you could write


MyModule.PrintThing(MyModule.MakeThing(firstThing, 76))

**Reply by dukester** [reply]
I see! Thx again ...

**Reply by dukester** [reply]
Worked like a charm! I'm surprised at, and appreciate the "skinny" executable the cm3 seems to cranks out.
